internal/gke/purge.go

```diff
@@ -0,0 +1,791 @@
+package gke
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	sync "sync"
+
+	artifactregistry "cloud.google.com/go/artifactregistry/apiv1beta2"
+	"cloud.google.com/go/artifactregistry/apiv1beta2/artifactregistrypb"
+	compute "cloud.google.com/go/compute/apiv1"
+	"cloud.google.com/go/compute/apiv1/computepb"
+	container "cloud.google.com/go/container/apiv1"
+	"cloud.google.com/go/container/apiv1/containerpb"
+	privateca "cloud.google.com/go/security/privateca/apiv1"
+	"cloud.google.com/go/security/privateca/apiv1/privatecapb"
+	"google.golang.org/api/dns/v1"
+	"google.golang.org/api/iam/v1"
+	"google.golang.org/api/iterator"
+)
+
+// Purge deletes all resources created by Service Weaver in the given GCP
+// project. It returns an error if there are some resourced that cannot
+// be deleted.
+// TODO: Update IAM permissions to not point to the deleted resources.
+func Purge(ctx context.Context, config CloudConfig) error {
+	var mu sync.Mutex
+	var errs []error
+	var waits []func() error
+	join := func() {
+		if len(waits) == 0 {
+			return
+		}
+		var wg sync.WaitGroup
+		wg.Add(len(waits))
+		for _, wait := range waits {
+			wait := wait
+			go func() {
+				defer wg.Done()
+				if err := wait(); err != nil {
+					mu.Lock()
+					errs = append(errs, err)
+					mu.Unlock()
+				}
+			}()
+		}
+		wg.Wait()
+		waits = waits[:0]
+	}
+
+	// Start deleting the artifacts repository, if necessary.
+	if wait, err := deleteRepository(ctx, config); err != nil {
+		errs = append(errs, err)
+	} else if waits != nil {
+		waits = append(waits, wait)
+	}
+
+	// Start deleting the clusters, if necessary.
+	if wait, err := deleteConfigCluster(ctx, config); err != nil {
+		errs = append(errs, err)
+	} else if wait != nil {
+		waits = append(waits, wait)
+	}
+	appClusters, err := getApplicationClusters(ctx, config)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	for _, cluster := range appClusters {
```

Comment:
I think we shouldn't touch `appClusters` if `getApplicationClusters` returns a non-nil error?